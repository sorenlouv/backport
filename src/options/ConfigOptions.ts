import { Logger } from '../services/logger';

export interface BranchChoice {
  name: string;
  checked?: boolean;
}
export type BranchChoiceRaw = string | BranchChoice;

type AutoFixConflictsHandler = ({
  files,
  directory,
  logger,
  targetBranch,
}: {
  files: string[];
  directory: string;
  logger: Logger;
  targetBranch: string;
}) => boolean | Promise<boolean>;

export type ConfigOptions = Partial<{
  accessToken: string;
  all: boolean;
  assignees: string[];
  author: string;
  autoAssign: boolean;
  autoFixConflicts: AutoFixConflictsHandler;
  branchLabelMapping: Record<string, string>;
  ci: boolean;
  editor: string;
  forceLocalConfig: boolean;
  fork: boolean;
  gitHostname: string;
  githubApiBaseUrlV3: string;
  githubApiBaseUrlV4: string;
  maxNumber: number;
  multiple: boolean;
  multipleBranches: boolean;
  multipleCommits: boolean;
  noVerify: boolean;
  path: string;
  prDescription: string;
  prFilter: string;
  prTitle: string;
  pullNumber: number;
  resetAuthor: boolean;
  sha: string;
  sourceBranch: string;
  sourcePRLabels: string[];
  targetBranchChoices: BranchChoiceRaw[];
  targetBranches: string[];
  targetPRLabels: string[];
  upstream: string;
  username: string;
  verbose: boolean;

  // deprecated: renamed to `targetBranchChoices`
  branches: BranchChoiceRaw[];

  // deprecated: renamed to `targetPRLabels`
  labels: string[];
}>;
