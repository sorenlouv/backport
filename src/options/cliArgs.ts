import yargs from 'yargs';
import { OptionsFromConfigFiles } from './config/config';

export type OptionsFromCliArgs = ReturnType<typeof getOptionsFromCliArgs>;
export function getOptionsFromCliArgs(
  configOptions: OptionsFromConfigFiles,
  argv: readonly string[]
) {
  const cliArgs = yargs(argv)
    .parserConfiguration({
      'strip-dashed': true,
      'strip-aliased': true,
      'boolean-negation': false,
    })
    .usage('$0 [args]')
    .wrap(Math.max(100, Math.min(120, yargs.terminalWidth())))
    .option('accessToken', {
      alias: 'accesstoken',
      description: 'Github access token',
      type: 'string',
    })
    .option('all', {
      default: configOptions.all,
      description: 'List all commits',
      alias: 'a',
      type: 'boolean',
    })
    .option('author', {
      default: configOptions.author,
      description: 'Show commits by specific author',
      type: 'string',
    })
    .option('maxNumber', {
      default: configOptions.maxNumber,
      description: 'Number of commits to choose from',
      alias: ['number', 'n'],
      type: 'number',
    })
    .option('dryRun', {
      default: false,
      description: 'Perform backport without pushing to Github',
      type: 'boolean',
    })
    .option('editor', {
      default: configOptions.editor,
      description: 'Editor to be opened during conflict resolution',
      type: 'string',
    })
    .option('fork', {
      default: configOptions.fork,
      description: 'Create backports in fork or origin repo',
      type: 'boolean',
    })
    .option('gitHostname', {
      hidden: true,
      default: configOptions.gitHostname,
      description: 'Hostname for Github',
      type: 'string',
    })
    .option('githubApiBaseUrlV3', {
      hidden: true,
      default: configOptions.githubApiBaseUrlV3,
      description: `Base url for Github's REST (v3) API`,
      type: 'string',
    })
    .option('githubApiBaseUrlV4', {
      hidden: true,
      default: configOptions.githubApiBaseUrlV4,
      description: `Base url for Github's GraphQL (v4) API`,
      type: 'string',
    })
    .option('mainline', {
      description:
        'Parent id of merge commit. Defaults to 1 when supplied without arguments',
      type: 'number',
      coerce: (mainline) => {
        // `--mainline` (default to 1 when no parent is given)
        if (mainline === undefined) {
          return 1;
        }

        // use specified mainline parent
        if (Number.isInteger(mainline)) {
          return mainline as number;
        }

        // Invalid value provided
        throw new Error(`--mainline must be an integer. Received: ${mainline}`);
      },
    })
    .option('multiple', {
      default: configOptions.multiple,
      description: 'Select multiple branches/commits',
      type: 'boolean',
    })
    .option('multipleCommits', {
      default: configOptions.multipleCommits,
      description: 'Backport multiple commits',
      type: 'boolean',
    })
    .option('multipleBranches', {
      default: configOptions.multipleBranches,
      description: 'Backport to multiple branches',
      type: 'boolean',
    })
    .option('noVerify', {
      default: configOptions.noVerify,
      description: 'Bypasses the pre-commit and commit-msg hooks',
      type: 'boolean',
    })
    .option('path', {
      default: configOptions.path,
      description: 'Only list commits touching files under the specified path',
      alias: 'p',
      type: 'string',
    })
    .option('prTitle', {
      default: configOptions.prTitle,
      description: 'Title of pull request',
      alias: 'title',
      type: 'string',
    })
    .option('prDescription', {
      default: configOptions.prDescription,
      description: 'Description to be added to pull request',
      alias: 'description',
      type: 'string',
    })
    .option('pullNumber', {
      conflicts: ['sha', 'sourcePRsFilter'],
      description: 'Pull request to backport',
      type: 'number',
      alias: 'pr',
    })
    .option('resetAuthor', {
      default: false,
      description: 'Set yourself as commit author',
      type: 'boolean',
    })
    .option('sha', {
      conflicts: ['pullNumber', 'sourcePRsFilter'],
      description: 'Commit sha to backport',
      type: 'string',
      alias: 'commit',
    })
    .option('sourcePRLabels', {
      default: configOptions.sourcePRLabels,
      description: 'Add labels to the source (original) PR',
      type: 'array',
      alias: 'sourcePRLabel',
    })
    .option('sourcePRsFilter', {
      conflicts: ['pullNumber', 'sha'],
      // default: configOptions.githubApiBaseUrlV4,
      description: `Filter source pull requests by a query`,
      alias: 'pr-filter',
      type: 'string',
    })
    .option('sourceBranch', {
      default: configOptions.sourceBranch,
      description: `List commits to backport from another branch than master`,
      type: 'string',
    })
    .option('verify', {
      description: `Opposite of no-verify`,
      type: 'boolean',
    })
    .option('targetBranches', {
      default: [] as string[],
      description: 'Branch(es) to backport to',
      type: 'array',
      alias: ['targetBranch', 'branch', 'b'],
      string: true, // ensure `6.0` is not coerced to `6`
    })
    .option('targetPRLabels', {
      default: configOptions.targetPRLabels,
      description: 'Add labels to the target (backport) PR',
      alias: ['labels', 'label', 'l'],
      type: 'array',
    })
    .option('upstream', {
      default: configOptions.upstream,
      description: 'Name of repository',
      alias: 'up',
      type: 'string',
    })
    .option('username', {
      default: configOptions.username,
      description: 'Github username',
      type: 'string',
    })
    .option('verbose', {
      default: false,
      description: 'Show additional debug information',
      type: 'boolean',
    })
    .alias('version', 'v')
    .alias('version', 'V')
    .help()
    .epilogue(
      'For bugs, feature requests or questions: https://github.com/sqren/backport/issues\nOr contact me directly: https://twitter.com/sorenlouv'
    ).argv;

  // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-unused-vars
  const { $0, _, verify, ...rest } = cliArgs;

  return {
    ...rest,
    accessToken: cliArgs.accessToken || configOptions.accessToken, // accessToken should not be displayed in yargs help menu
    branchLabelMapping: configOptions.branchLabelMapping, // not available as cli argument
    multipleBranches: cliArgs.multipleBranches || cliArgs.multiple,
    multipleCommits: cliArgs.multipleCommits || cliArgs.multiple,
    noVerify: verify ?? rest.noVerify, // `verify` is a cli-only flag to flip the default of `no-verify`
    targetBranchChoices: configOptions.targetBranchChoices, // not available as cli argument
  };
}
