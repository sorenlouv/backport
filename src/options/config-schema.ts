import { z } from 'zod';
import type { TargetBranchChoiceOrString } from './config-options';

// Target branch choice schema
const targetBranchChoiceSchema: z.ZodType<TargetBranchChoiceOrString> = z.union(
  [
    z.string(),
    z.object({
      name: z.string(),
      checked: z.boolean().optional(),
    }),
  ],
);

// Base config schema WITHOUT defaults - used for parsing
// Defaults are applied later in the merging process
const baseConfigSchema = z.object({
  accessToken: z.string().optional(),
  assignees: z.array(z.string()).optional(),
  author: z.union([z.string(), z.null()]).optional(),
  autoAssign: z.boolean().optional(),
  autoMerge: z.boolean().optional(),
  autoMergeMethod: z.enum(['merge', 'rebase', 'squash']).optional(),
  backportBinary: z.string().optional(),
  backportBranchName: z.string().optional(),
  cherrypickRef: z.boolean().optional(),
  commitConflicts: z.boolean().optional(),
  commitPaths: z.array(z.string()).optional(),
  copySourcePRLabels: z.boolean().optional(),
  copySourcePRReviewers: z.boolean().optional(),
  details: z.boolean().optional(),
  dir: z.string().optional(),
  draft: z.boolean().optional(),
  dryRun: z.boolean().optional(),
  editor: z.string().optional(),
  fork: z.boolean().optional(),
  gitAuthorEmail: z.string().optional(),
  gitAuthorName: z.string().optional(),
  gitHostname: z.string().optional(),
  githubActionRunId: z.number().optional(),
  githubApiBaseUrlV3: z.string().optional(),
  githubApiBaseUrlV4: z.string().optional(),
  interactive: z.boolean().optional(),
  logFilePath: z.string().optional(),
  ls: z.boolean().optional(),
  maxNumber: z.number().optional(),
  mainline: z.number().optional(),
  multiple: z.boolean().optional(),
  multipleBranches: z.boolean().optional(),
  multipleCommits: z.boolean().optional(),
  noVerify: z.boolean().optional(),
  prDescription: z.string().optional(),
  prFilter: z.string().optional(),
  projectConfigFile: z.string().optional(),
  prTitle: z.string().optional(),
  publishStatusCommentOnAbort: z.boolean().optional(),
  publishStatusCommentOnFailure: z.boolean().optional(),
  publishStatusCommentOnSuccess: z.boolean().optional(),
  pullNumber: z.union([z.number(), z.array(z.number())]).optional(),
  repoForkOwner: z.string().optional(),
  repoName: z.string().optional(),
  repoOwner: z.string().optional(),
  resetAuthor: z.boolean().optional(),
  reviewers: z.array(z.string()).optional(),
  sha: z.union([z.string(), z.array(z.string())]).optional(),
  signoff: z.boolean().optional(),
  skipRemoteConfig: z.boolean().optional(),
  noUnmergedBackportsHelp: z.boolean().optional(),
  sourceBranch: z.string().optional(),
  sourcePRLabels: z.array(z.string()).optional(),
  targetBranchChoices: z.array(targetBranchChoiceSchema).optional(),
  targetBranches: z.array(z.string()).optional(),
  targetPRLabels: z.array(z.string()).optional(),
  telemetry: z.boolean().optional(),
  cwd: z.string().optional(),
  dateSince: z.string().nullable().optional(),
  dateUntil: z.string().nullable().optional(),
  globalConfigFile: z.string().optional(),

  // Deprecated options (for backward compatibility)
  upstream: z.string().optional(),
  branches: z.array(targetBranchChoiceSchema).optional(),
  labels: z.array(z.string()).optional(),
  addOriginalReviewers: z.boolean().optional(),

  // Config file only options
  branchLabelMapping: z.record(z.string(), z.string()).optional(),

  // yargs-only options (for parsing, not for validation)
  help: z.boolean().optional(),
  version: z.boolean().optional(),
  v: z.boolean().optional(),
});

// Schema for partial config (from files/cli) - without defaults applied
// This is used for parsing CLI args and config files where we don't want defaults
// Defaults are applied later via manual merge in options.ts
export const partialConfigSchema = baseConfigSchema.loose();
