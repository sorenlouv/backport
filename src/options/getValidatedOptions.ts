import isEmpty from 'lodash.isempty';
import { HandledError } from '../services/HandledError';
import { getGlobalConfigPath } from '../services/env';
import { OptionsFromCliArgs } from './cliArgs';

const GLOBAL_CONFIG_DOCS_LINK =
  'https://github.com/sqren/backport/blob/e119d71d6dc03cd061f6ad9b9a8b1cd995f98961/docs/configuration.md#global-config-backportconfigjson';

const PROJECT_CONFIG_DOCS_LINK =
  'https://github.com/sqren/backport/blob/e119d71d6dc03cd061f6ad9b9a8b1cd995f98961/docs/configuration.md#project-config-backportrcjson';

export type ValidatedOptions = ReturnType<typeof getValidatedOptions>;
export function getValidatedOptions({
  // omit `upstream` from options
  upstream,
  ...options
}: OptionsFromCliArgs) {
  // ensure accessToken and username are given
  if (!options.accessToken || !options.username) {
    const globalConfigPath = getGlobalConfigPath();
    throw new HandledError(
      `Please update your config file: ${globalConfigPath}.\nIt must contain a valid "username" and "accessToken".\n\nRead more: ${GLOBAL_CONFIG_DOCS_LINK}`
    );
  }

  // ensure targetBranches or targetBranchChoices are given
  if (isEmpty(options.targetBranches) && isEmpty(options.targetBranchChoices)) {
    throw new HandledError(
      `You must specify a target branch\n\nYou can specify it via either:\n - Config file (recommended): ".backportrc.json". Read more: ${PROJECT_CONFIG_DOCS_LINK}\n - CLI: "--branch 6.1"`
    );
  }

  // ensure a valid upstream is given
  const [repoOwner, repoName] = (upstream ?? '').split('/');
  if (!repoOwner || !repoName) {
    throw new HandledError(
      `You must specify a valid Github repository\n\nYou can specify it via either:\n - Config file (recommended): ".backportrc.json". Read more: ${PROJECT_CONFIG_DOCS_LINK}\n - CLI: "--upstream elastic/kibana"`
    );
  }

  return {
    ...options,

    // no longer optional
    accessToken: options.accessToken,
    username: options.username,

    // split upstream
    repoName,
    repoOwner,

    // define author
    author: options.author || options.username,
    all: options.author ? false : options.all,
  };
}
