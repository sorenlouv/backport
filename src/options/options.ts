import isEmpty from 'lodash.isempty';
import { HandledError } from '../services/HandledError';
import { OptionsFromCliArgs, getOptionsFromCliArgs } from './cliArgs';
import { getOptionsFromConfigFiles } from './config/config';
import { PromiseReturnType } from '../types/commons';
import { getGlobalConfigPath } from '../services/env';

export type BackportOptions = PromiseReturnType<typeof getOptions>;
export async function getOptions(argv: typeof process.argv) {
  const optionsFromConfig = await getOptionsFromConfigFiles();
  const optionsFromCli = getOptionsFromCliArgs(optionsFromConfig, argv);
  return validateOptions(optionsFromCli);
}

const GLOBAL_CONFIG_DOCS_LINK =
  'https://github.com/sqren/backport/blob/b9e2ec8e7a56bcbee96a8e9d38088ddd6ccaf65b/docs/configuration.md#userglobal-configuration';

const PROJECT_CONFIG_DOCS_LINK =
  'https://github.com/sqren/backport/blob/b9e2ec8e7a56bcbee96a8e9d38088ddd6ccaf65b/docs/configuration.md#project-specific-configuration';

function getErrorMessage({
  field,
  exampleValue
}: {
  field: keyof OptionsFromCliArgs;
  exampleValue: string;
}) {
  const isGlobalConfigProperty =
    field === 'accessToken' || field === 'username';

  const globalConfigPath = getGlobalConfigPath();

  const configFileMessage = isGlobalConfigProperty
    ? `Config file: "${globalConfigPath}". Read more: ${GLOBAL_CONFIG_DOCS_LINK}`
    : `Config file: ".backportrc.json". Read more: ${PROJECT_CONFIG_DOCS_LINK}`;

  return `Invalid option "${field}"\n\nYou can add it with either:\n - ${configFileMessage}\n - CLI: "--${field} ${exampleValue}"`;
}

export function validateOptions({
  accessToken,
  all,
  branchChoices,
  branches,
  labels,
  multiple,
  multipleBranches,
  multipleCommits,
  sha,
  upstream,
  username
}: OptionsFromCliArgs) {
  if (!accessToken) {
    throw new HandledError(
      getErrorMessage({ field: 'accessToken', exampleValue: 'myAccessToken' })
    );
  }

  if (isEmpty(branches) && isEmpty(branchChoices)) {
    throw new HandledError(
      getErrorMessage({ field: 'branches', exampleValue: '6.1' })
    );
  }

  if (!upstream) {
    throw new HandledError(
      getErrorMessage({ field: 'upstream', exampleValue: 'elastic/kibana' })
    );
  }

  if (!username) {
    throw new HandledError(
      getErrorMessage({ field: 'username', exampleValue: 'sqren' })
    );
  }

  return {
    accessToken,
    all,
    branchChoices,
    branches,
    labels,
    multiple,
    multipleBranches,
    multipleCommits,
    sha,
    upstream,
    username
  };
}
