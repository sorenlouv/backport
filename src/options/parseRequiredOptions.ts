import isEmpty from 'lodash.isempty';
import isString from 'lodash.isstring';
import { HandledError } from '../services/HandledError';
import { getGlobalConfigPath } from '../services/env';
import { OptionsFromGithub } from '../services/github/v4/getOptionsFromGithub';
import { BranchChoiceRaw, BranchChoice, ConfigOptions } from './ConfigOptions';
import { OptionsFromCliArgs } from './cliArgs';
import { OptionsFromConfigFiles } from './config/config';

const PROJECT_CONFIG_DOCS_LINK =
  'https://github.com/sqren/backport/blob/e119d71d6dc03cd061f6ad9b9a8b1cd995f98961/docs/configuration.md#project-config-backportrcjson';

const GLOBAL_CONFIG_DOCS_LINK =
  'https://github.com/sqren/backport/blob/e119d71d6dc03cd061f6ad9b9a8b1cd995f98961/docs/configuration.md#global-config-backportconfigjson';

export type ValidatedOptions = ReturnType<typeof parseRequiredOptions>;
export function parseRequiredOptions(
  optionsFromConfigFiles: OptionsFromConfigFiles,
  optionsFromCliArgs: OptionsFromCliArgs,
  optionsFromGithub: OptionsFromGithub
) {
  const options = {
    // local config files have lowest precedence
    ...optionsFromConfigFiles,

    // remote config options override local config options
    ...optionsFromGithub,

    // cli args override both of the above
    ...optionsFromCliArgs,
  } as OptionsFromCliArgs & OptionsFromConfigFiles & OptionsFromGithub;

  // ensure `targetBranches` or `targetBranchChoices` are given
  if (
    isEmpty(options.targetBranches) &&
    isEmpty(options.targetBranchChoices) &&
    // this is primarily necessary on CI where `targetBranches` and `targetBranchChoices` and not given
    isEmpty(options.branchLabelMapping)
  ) {
    throw new HandledError(
      `You must specify a target branch\n\nYou can specify it via either:\n - Config file (recommended): ".backportrc.json". Read more: ${PROJECT_CONFIG_DOCS_LINK}\n - CLI: "--branch 6.1"`
    );
  }

  const { accessToken, username, repoName, repoOwner } = getRequiredOptions(
    options
  );

  return {
    ...options,

    // convert from array of primitives to array of object
    targetBranchChoices: getTargetBranchChoicesAsObject(
      options.targetBranchChoices
    ),

    // required options
    accessToken,
    username,
    repoName,
    repoOwner,

    // auto-assign the current user or default to the specified assignees
    assignees: options.autoAssign
      ? [options.username as string]
      : options.assignees,

    author: options.author || username,
    all: options.author ? false : options.all,
  };
}

// `branches` can either be a string or an object.
// It must be transformed so it is always treated as an object troughout the application
function getTargetBranchChoicesAsObject(
  targetBranchChoices: BranchChoiceRaw[]
): BranchChoice[] {
  return targetBranchChoices.map((choice) => {
    if (isString(choice)) {
      return {
        name: choice,
        checked: false,
      };
    }

    return choice;
  });
}

export function getRequiredOptions({
  accessToken,
  username,
  upstream,
}: ConfigOptions) {
  // accessToken and username must be supplied in config or via cli args
  if (!accessToken || !username) {
    const globalConfigPath = getGlobalConfigPath();
    throw new HandledError(
      `Please update your config file: ${globalConfigPath}.\nIt must contain a valid "username" and "accessToken".\n\nRead more: ${GLOBAL_CONFIG_DOCS_LINK}`
    );
  }

  // upstream must be specified via config or cli args
  const [repoOwner, repoName] = (upstream ?? '').split('/');
  if (!repoOwner || !repoName) {
    throw new HandledError(
      `You must specify a valid Github repository\n\nYou can specify it via either:\n - Config file (recommended): ".backportrc.json". Read more: ${PROJECT_CONFIG_DOCS_LINK}\n - CLI: "--upstream elastic/kibana"`
    );
  }

  return { accessToken, username, repoOwner, repoName };
}
