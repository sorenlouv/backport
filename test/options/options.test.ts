import { validateOptions } from '../../src/options/options';

const validOptions = {
  accessToken: 'myAccessToken',
  all: false,
  branchChoices: [],
  branches: ['branchA'],
  labels: [],
  multiple: false,
  multipleBranches: true,
  multipleCommits: false,
  sha: undefined,
  upstream: 'elastic/kibana',
  username: 'sqren'
};

describe('validateOptions', () => {
  describe('should not throw', () => {
    it('when all options are valid and `branches` is given', () => {
      expect(() => validateOptions(validOptions)).not.toThrow();
    });

    it('when all options are valid and `branchChoices` is given', () => {
      expect(() =>
        validateOptions({
          ...validOptions,
          branchChoices: [{ name: 'branchA' }],
          branches: []
        })
      ).not.toThrow();
    });
  });

  describe('should throw', () => {
    it('when accessToken is missing', () => {
      expect(() =>
        validateOptions({
          ...validOptions,
          accessToken: undefined
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when accessToken is empty', () => {
      expect(() =>
        validateOptions({
          ...validOptions,
          accessToken: ''
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when both branches and branchChoices are missing', () => {
      expect(() =>
        validateOptions({
          ...validOptions,
          branchChoices: [],
          branches: []
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when upstream is missing', () => {
      expect(() =>
        validateOptions({
          ...validOptions,
          upstream: undefined
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when upstream is empty', () => {
      expect(() =>
        validateOptions({
          ...validOptions,
          upstream: ''
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when username is missing', () => {
      expect(() =>
        validateOptions({
          ...validOptions,
          username: undefined
        })
      ).toThrowErrorMatchingSnapshot();
    });

    it('when username is empty', () => {
      expect(() =>
        validateOptions({
          ...validOptions,
          username: ''
        })
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
