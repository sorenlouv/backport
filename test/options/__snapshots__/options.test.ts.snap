// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateOptions should throw when accessToken is empty 1`] = `
"Invalid option \\"accessToken\\"

You can add it with either:
 - Config file: \\"/myHomeDir/.backport/config.json\\". Read more: https://github.com/sqren/backport/blob/b9e2ec8e7a56bcbee96a8e9d38088ddd6ccaf65b/docs/configuration.md#userglobal-configuration
 - CLI: \\"--accessToken myAccessToken\\""
`;

exports[`validateOptions should throw when accessToken is missing 1`] = `
"Invalid option \\"accessToken\\"

You can add it with either:
 - Config file: \\"/myHomeDir/.backport/config.json\\". Read more: https://github.com/sqren/backport/blob/b9e2ec8e7a56bcbee96a8e9d38088ddd6ccaf65b/docs/configuration.md#userglobal-configuration
 - CLI: \\"--accessToken myAccessToken\\""
`;

exports[`validateOptions should throw when both branches and branchChoices are missing 1`] = `
"Invalid option \\"branches\\"

You can add it with either:
 - Config file: \\".backportrc.json\\". Read more: https://github.com/sqren/backport/blob/b9e2ec8e7a56bcbee96a8e9d38088ddd6ccaf65b/docs/configuration.md#project-specific-configuration
 - CLI: \\"--branches 6.1\\""
`;

exports[`validateOptions should throw when upstream is empty 1`] = `
"Invalid option \\"upstream\\"

You can add it with either:
 - Config file: \\".backportrc.json\\". Read more: https://github.com/sqren/backport/blob/b9e2ec8e7a56bcbee96a8e9d38088ddd6ccaf65b/docs/configuration.md#project-specific-configuration
 - CLI: \\"--upstream elastic/kibana\\""
`;

exports[`validateOptions should throw when upstream is missing 1`] = `
"Invalid option \\"upstream\\"

You can add it with either:
 - Config file: \\".backportrc.json\\". Read more: https://github.com/sqren/backport/blob/b9e2ec8e7a56bcbee96a8e9d38088ddd6ccaf65b/docs/configuration.md#project-specific-configuration
 - CLI: \\"--upstream elastic/kibana\\""
`;

exports[`validateOptions should throw when username is empty 1`] = `
"Invalid option \\"username\\"

You can add it with either:
 - Config file: \\"/myHomeDir/.backport/config.json\\". Read more: https://github.com/sqren/backport/blob/b9e2ec8e7a56bcbee96a8e9d38088ddd6ccaf65b/docs/configuration.md#userglobal-configuration
 - CLI: \\"--username sqren\\""
`;

exports[`validateOptions should throw when username is missing 1`] = `
"Invalid option \\"username\\"

You can add it with either:
 - Config file: \\"/myHomeDir/.backport/config.json\\". Read more: https://github.com/sqren/backport/blob/b9e2ec8e7a56bcbee96a8e9d38088ddd6ccaf65b/docs/configuration.md#userglobal-configuration
 - CLI: \\"--username sqren\\""
`;
